sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/SqlToRelCompiler.java

```diff
@@ -734,6 +734,65 @@ public RelDataType specToRel(SqlDataTypeSpec spec, boolean neverNullable) {
         return result;
     }
 
+    /** True if the left data type can be cast to the right data type "trivially".
+     * This includes:
+     * - non-null to null
+     * - char(n) to varchar */
+    @SuppressWarnings("RedundantIfStatement")
+    public static boolean canBeTriviallyCastTo(RelDataType left, RelDataType right) {
+        if (left.equals(right))
+            return true;
+        SqlTypeName leftName = left.getSqlTypeName();
+        SqlTypeName rightName = right.getSqlTypeName();
+        if (left.isNullable() && !right.isNullable())
+            return false;
+        if (leftName != rightName) {
+            if (leftName == SqlTypeName.CHAR && rightName == SqlTypeName.VARCHAR) {
+                if (right.getPrecision() == RelDataType.PRECISION_NOT_SPECIFIED) {
+                    return true;
+                }
+            } else if (leftName == SqlTypeName.NULL && right.isNullable()) {
+                return true;
+            } else if (SqlTypeName.INT_TYPES.contains(leftName) && SqlTypeName.INT_TYPES.contains(rightName)) {
+                if (left.getPrecision() <= right.getPrecision())
+                    return true;
+            }
+            return false;
+        }
+        if (left.isStruct() != right.isStruct())
+            return false;
+        if (left.isStruct()) {
+            if (left.getFieldCount() != right.getFieldCount())
+                return false;
+            List<RelDataTypeField> leftFields = left.getFieldList();
+            List<RelDataTypeField> rightFields = right.getFieldList();
+            for (int i = 0; i < left.getFieldCount(); i++) {
+                RelDataTypeField lf = leftFields.get(i);
+                RelDataTypeField rf = rightFields.get(i);
+                if (!lf.getName().equals(rf.getName()))
+                    return false;
+
+                RelDataType lt = lf.getType();
+                RelDataType rt = rf.getType();
+                if (!canBeTriviallyCastTo(lt, rt))
+                    return false;
+            }
+        } else if (left.getComponentType() != null) {
```

Comment:
I assume this is for arrays?

Reply:
and multisets, but we'll reject these someplace else.
maybe someday we'll support them